libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -431,48 +431,106 @@ void GCS_MAVLINK::send_ahrs3()
 #endif
 }
 
-/*
-  handle a MISSION_REQUEST_LIST mavlink packet
- */
+// handle a request for the number of items we have stored for a mission type:
 void GCS_MAVLINK::handle_mission_request_list(AP_Mission &mission, mavlink_message_t *msg)
 {
     // decode
     mavlink_mission_request_list_t packet;
     mavlink_msg_mission_request_list_decode(msg, &packet);
 
-    // reply with number of commands in the mission.  The GCS will then request each command separately
-    mavlink_msg_mission_count_send(chan,msg->sysid, msg->compid, mission.num_commands(),
-                                   MAV_MISSION_TYPE_MISSION);
+    uint16_t count;
+    switch (packet.mission_type) {
+    case MAV_MISSION_TYPE_MISSION:
+        count = mission.num_commands();
+        waypoint_receiving = false;             // record that we are sending commands (i.e. not receiving)
+        break;
+    default:
+        // should we be sending a MISSION_ACK here with a failure code?
+        return;
+    }
 
-    // set variables to help handle the expected sending of commands to the GCS
-    waypoint_receiving = false;             // record that we are sending commands (i.e. not receiving)
+    // reply with number of commands in the mission.  The GCS will
+    // then request each command separately
+    mavlink_msg_mission_count_send(chan,
+                                   msg->sysid,
+                                   msg->compid,
+                                   count,
+                                   packet.mission_type);
 }
 
 /*
   handle a MISSION_REQUEST mavlink packet
  */
-void GCS_MAVLINK::handle_mission_request(AP_Mission &mission, mavlink_message_t *msg)
+void GCS_MAVLINK::handle_mission_request_int(mavlink_message_t *msg)
 {
     AP_Mission::Mission_Command cmd;
 
-    if (msg->msgid == MAVLINK_MSG_ID_MISSION_REQUEST_INT) {  
         // decode
         mavlink_mission_request_int_t packet;
         mavlink_msg_mission_request_int_decode(msg, &packet);
 
-        // retrieve mission from eeprom
-        if (!mission.read_cmd_from_storage(packet.seq, cmd)) {
+        mavlink_mission_item_int_t ret_packet{};
+
+        ret_packet.target_system = msg->sysid;
+        ret_packet.target_component = msg->compid;
+        ret_packet.seq = packet.seq;
+        ret_packet.mission_type = packet.mission_type;
+
+        MAV_MISSION_RESULT failure_code = MAV_MISSION_ERROR;
+
+        switch (packet.mission_type) {
+        case MAV_MISSION_TYPE_MISSION:
+            if (!handle_mission_request_int_mission(packet, ret_packet)) {
```

Comment:
It seems unfortunate that this can't provide a failure code. We actually have meaningful internal reasons for failing rather then just `ERROR`. (`UNSUPPORTED` and `INVALID_SEQUENCE` being the main ones that come to mind)

Reply:
You're thinking we could return a `MAV_MISSION_RESULT`?

Reply:
Pushed up a patch to do that.